File: task2/test_api.js
async function testAPI() {
    try {
        // Test API response
        const response = await fetch('https://fakestoreapi.com/products');
        
        console.log('1. Testing API response code:');
        console.log(`Status code: ${response.status}`);
        console.log(response.status === 200 ? '✅ Success' : '❌ Failed');
        console.log('-'.repeat(50));

        // Get products data
        const products = await response.json();
        const defects = [];

        console.log('2. Testing products data:');
        products.forEach(product => {
            const productDefects = [];

            // Check title
            if (!product.title || product.title.trim() === '') {
                productDefects.push('Empty title');
            }

            // Check price
            if (product.price === undefined || product.price < 0) {
                productDefects.push('Invalid price (negative or missing)');
            }

            // Check rating
            if (!product.rating?.rate || product.rating.rate > 5) {
                productDefects.push('Invalid rating (exceeds 5 or missing)');
            }

            if (productDefects.length > 0) {
                defects.push({
                    id: product.id,
                    title: product.title,
                    defects: productDefects
                });
            }
        });

        // Display results
        if (defects.length === 0) {
            console.log('✅ All products passed validation');
        } else {
            console.log('❌ Found products with defects:');
            defects.forEach(item => {
                console.log(`\nProduct ID: ${item.id}`);
                console.log(`Title: ${item.title}`);
                console.log(`Defects: ${item.defects.join(', ')}`);
            });
        }
        console.log('-'.repeat(50));
        console.log(`Total products tested: ${products.length}`);
        console.log(`Products with defects: ${defects.length}`);

    } catch (error) {
        console.error('Error during testing:', error.message);
    }
}

// Run the tests
testAPI();

File: task2/test_api.py
import requests
import pytest

API_URL = "https://fakestoreapi.com/products"

def test_api_response():
    """Test API response status code"""
    response = requests.get(API_URL)
    assert response.status_code == 200, f"Expected status code 200, got {response.status_code}"

def test_product_validation():
    """Test product data validation and collect defects"""
    response = requests.get(API_URL)
    products = response.json()
    defects = []

    for product in products:
        product_defects = []
        
        # Check title
        if not product.get('title') or len(product['title'].strip()) == 0:
            product_defects.append("Empty title")

        # Check price
        price = product.get('price')
        if price is None or price < 0:
            product_defects.append("Invalid price (negative or missing)")

        # Check rating
        rating = product.get('rating', {}).get('rate')
        if rating is None or rating > 5:
            product_defects.append("Invalid rating (exceeds 5 or missing)")

        if product_defects:
            defects.append({
                'id': product.get('id'),
                'title': product.get('title'),
                'defects': product_defects
            })

    # Print defects report if any found
    if defects:
        print("\nProducts with defects:")
        for item in defects:
            print(f"Product ID: {item['id']}")
            print(f"Title: {item['title']}")
            print(f"Defects: {', '.join(item['defects'])}")
            print("-" * 50)

    # Assert no defects were found
    assert len(defects) == 0, f"Found {len(defects)} products with defects"

if __name__ == '__main__':
    pytest.main([__file__, "-v"])
